[mlir][OpenMP] - Transform target offloading directives for easier translation to LLVMIR

bhandarkar-pranav opened the pull request:
This patch adds a pass that transforms `omp.target`, `omp.target_enter_data`,
  `omp.target_update_data` and `omp.target_exit_data` whenever these operations
  have the `depend` clause on them.

  These operations are transformed by enclosing them inside a new `omp.task`
  operation. The `depend` clause related arguments are moved to the new `omp.task`
  operation from the original 'target' operation.

  Example:
  Input:
  ```mlir
  omp.target map_entries(%map_a -> %arg0, %map_c_from -> %arg1 : memref<?xi32>, memref<?xi32>) depend(taskdependout -> %c : memref<?xi32>) {
    ^bb0(%arg0 : memref<?xi32>, %arg1 : memref<?xi32>) :
        "test.foobar"() : ()->()
         omp.terminator
  }
  ```
  Output:
  ```mlir
  omp.task depend(taskdependout -> %c : memref<?xi32>) {
    omp.target map_entries(%map_a -> %arg0, %map_c_from -> %arg1 : memref<?xi32>, memref<?xi32>) {
      ^bb0(%arg0 : memref<?xi32>, %arg1 : memref<?xi32>) :
          "test.foobar"() : ()->()
           omp.terminator
    }
    omp.terminator
  }
```
 The intent is to make it easier to translate to LLVMIR by avoiding the creation of such tasks in the `OMPIRBuilder`.

bhandarkar-pranav committed:
Checkpoint commit - Add buildDependData

bhandarkar-pranav committed:
checkpoint commit - pass openmp-task-based-target set up

bhandarkar-pranav committed:
Add patterns that match but do not rewrite anytyhing yet

bhandarkar-pranav committed:
Working for a simple testcase

bhandarkar-pranav committed:
Exten openmp-task-based-target transformation to target enter/update/exit data as well

bhandarkar-pranav committed:
Remove some debug prints

bhandarkar-pranav committed:
Fix top level comment in OpenMPTaskBasedTarget.cpp

bhandarkar-pranav committed:
Add a description for openmp-task-based-target pass

bhandarkar-pranav committed:
Fix the description of openmp-task-based-target pass

bhandarkar-pranav committed:
Roll back a now unnecessary change in mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

bhandarkar-pranav committed:
Roll back one more uninted change in mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

skatrak commented:
Is there anything preventing this from being done as part of PFT to MLIR lowering rather than a pass? It should be possible to make the `ClauseProcessor::processDepend` method create the outer `omp.task` with the corresponding DEPEND clause, and then have a simpler `ClauseProcessor::processTaskDepend` for the case when it's already attached to the TASK construct (so no new `omp.task` is created then). `omp.target*` ops wouldn't need any DEPEND-related arguments at that point.

In my opinion, that would be less expensive than adding a full pass, and in principle it shouldn't create any big compromises that I can think of. Do you agree with that? I don't mind this approach either, if it's generally preferred by others.

bhandarkar-pranav commented:
> Is there anything preventing this from being done as part of PFT to MLIR lowering rather than a pass? It should be possible to make the `ClauseProcessor::processDepend` method create the outer `omp.task` with the corresponding DEPEND clause, and then have a simpler `ClauseProcessor::processTaskDepend` for the case when it's already attached to the TASK construct (so no new `omp.task` is created then). `omp.target*` ops wouldn't need any DEPEND-related arguments at that point.
> 
> In my opinion, that would be less expensive than adding a full pass, and in principle it shouldn't create any big compromises that I can think of. Do you agree with that? I don't mind this approach either, if it's generally preferred by others.

Thank you for your reply @skatrak.
I don't think there is anything preventing it from being done in flang (PFT->MLIR lowering) and in fact, @ergawy had even suggested this in my discussion with him. However, my reasons for doing it this way largely boil down to preference for the following reasons (in no particular order)
- Keeping it in MLIR allows other frontends to use this because for any frontend to realize offloading in  LLVMIR, it'll have to go the route of creating an outer enclosing task
- This way, arguably, keeps the interface of PFT->MLIR lowering cleaner - There will only be one process method handling ``depend`` -> ``processDepend`` and all ops using the ``depend`` clause will have the same interface (aka ``processDepend``)
- It separates the concerns of lowering (PFT->MLIR) and codegen (this transformation is needed essentially for codegen)

I felt at the time that these were reasons enough for paying the price of going over the IR in a new pass altogether. Having said that, it isn't a very very strong preference at all and I am happy to change my approach should that be required. How about I leave this open for others to comment on, while I test this with an end-to-end testcase by invoking this pass from within flang?

ergawy commented on mlir/include/mlir/Dialect/OpenMP/Passes.td:
```diff
@@ -0,0 +1,54 @@
+//===-- Passes.td - OpenMP pass definition file -------------*- tablegen -*-===//
```
Previous OpenMP passes were added in https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Transforms/Passes.td#L321.

Even though I agree that OpenMP passes like this one and the ones linked above can be separated into their own file since they are not flang specific, but I think for consistency it would be better to use the already existing file to have a more centralized view of all OpenMP-related passes.

agozillon replied:
Just a little background on why the passes are there from what I recall at least (more for information to help make informed decisions or for someone else to chime in with a preference), the `OMPMarkDeclareTargetPass`/`OMPFunctionFiltering `are there as they were initially in an experimental phase (alongside another now erased pass) to verify if we'd keep them in the offloading flow and see how they'd interact with each other overtime as they are interlinked behavior. Not sure when they'd best to be moved, it's coming up to their anniversary at this point, but it's something that can wait in any case.

The `OMPDescriptorMapInfoGenPass `is very much Fortran specific at the moment, it primarily handles FIR `BoxType`'s, so it's not feasible at the moment (or perhaps ever) to move to the OpenMP dialect as it has FIR dependencies and FIR is an external dialect to the MLIR project.

bhandarkar-pranav replied:
Thanks for the info @agozillon. @ergawy  - Like you said, I kept this in mlir only because it isnt really flang specific and I anticipated that if I put it in flang someone might call me out for it. OTOH keeping it in flang would have made testcase generation easier as I'd have been able to load the `fir` dialect into the MLIR context. Now I had to rely on ``fir-opt`` to convert all ``fir`` ops to ``llvm`` dialect ops, but  `fir-opt` is broken in a weird way such that I couldn't run ``fir-to-llvm-ir`` lowering to rid my testcases of ``fir`` ops. (Aside: I hope to find some time to chase down the `fir-opt` problem)

ergawy commented on mlir/lib/Dialect/OpenMP/Transforms/OpenMPTaskBasedTarget.cpp:
```diff
@@ -0,0 +1,120 @@
+
+    // Only match a target op with a  'depend' clause on it.
+    if (op.getDependVars().empty()) {
+      return rewriter.notifyMatchFailure(op, "depend clause not found on op");
```
Why not use a `ConversionTarget` and setup it with legality information instead? https://github.com/llvm/llvm-project/blob/main/mlir/docs/DialectConversion.md#conversion-target

bhandarkar-pranav replied:
I did consider it and correct me if I am wrong, but a `ConversionPattern` is needed only if types need to be legalized as well correct? Since I didn't need that I chose to use an `OpRewritePattern`. Happy to change it if there is a strong preference. BTW, thanks for pointing me to your 'do concurrent' PR, that helped me start up quickly.

ergawy reviewed (commented):
Just a few small comments. I like the pass approach better than handling this `OpenMP.cpp` tbh :).

jsjodin commented:
> > Is there anything preventing this from being done as part of PFT to MLIR lowering rather than a pass? It should be possible to make the `ClauseProcessor::processDepend` method create the outer `omp.task` with the corresponding DEPEND clause, and then have a simpler `ClauseProcessor::processTaskDepend` for the case when it's already attached to the TASK construct (so no new `omp.task` is created then). `omp.target*` ops wouldn't need any DEPEND-related arguments at that point.
> > In my opinion, that would be less expensive than adding a full pass, and in principle it shouldn't create any big compromises that I can think of. Do you agree with that? I don't mind this approach either, if it's generally preferred by others.
> 
> Thank you for your reply @skatrak. I don't think there is anything preventing it from being done in flang (PFT->MLIR lowering) and in fact, @ergawy had even suggested this in my discussion with him. However, my reasons for doing it this way largely boil down to preference for the following reasons (in no particular order)
> 
> * Keeping it in MLIR allows other frontends to use this because for any frontend to realize offloading in  LLVMIR, it'll have to go the route of creating an outer enclosing task
> * This way, arguably, keeps the interface of PFT->MLIR lowering cleaner - There will only be one process method handling `depend` -> `processDepend` and all ops using the `depend` clause will have the same interface (aka `processDepend`)
> * It separates the concerns of lowering (PFT->MLIR) and codegen (this transformation is needed essentially for codegen)
> 
> I felt at the time that these were reasons enough for paying the price of going over the IR in a new pass altogether. Having said that, it isn't a very very strong preference at all and I am happy to change my approach should that be required. How about I leave this open for others to comment on, while I test this with an end-to-end testcase by invoking this pass from within flang?

It seems like this is similar to what @kparzysz is working on addressing, not quite the same perhaps, but it would be nice if we could produce MLIR directly that has the omp.task op. If we do that, would it be possible to remove the depend clause from the omp.target op?

bhandarkar-pranav committed:
Update to fix build and tests after rebasing on main.

Also add the if clause to the newly generated omp.task op that
encloses the omp.target op.

bhandarkar-pranav committed:
fix formatting issues
